api/PayrollProcessor.Data.Persistence/Features/Departments/DepartmentPayrollsQueryHandler.cs

```diff
@@ -0,0 +1,67 @@
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using Ardalis.GuardClauses;
+using LanguageExt;
+using Microsoft.Azure.Cosmos;
+using Microsoft.Azure.Cosmos.Linq;
+using PayrollProcessor.Core.Domain.Features.Departments;
+using PayrollProcessor.Core.Domain.Intrastructure.Operations.Queries;
+
+using static LanguageExt.Prelude;
+
+namespace PayrollProcessor.Data.Persistence.Features.Departments
+{
+    public class DepartmentPayrollsQueryHandler : IQueryHandler<DepartmentPayrollsQuery, IEnumerable<DepartmentPayroll>>
+    {
+        private readonly CosmosClient client;
+
+        public DepartmentPayrollsQueryHandler(CosmosClient client)
+        {
+            Guard.Against.Null(client, nameof(client));
+
+            this.client = client;
+        }
+
+        public TryOptionAsync<IEnumerable<DepartmentPayroll>> Execute(DepartmentPayrollsQuery query, CancellationToken token = default)
+        {
+            var (count, department, checkDateFrom, checkDateTo) = query;
+
+            var dataQuery = client
```

Comment:
I could put this 'setup' code inside the callback returned by this method. That would allow the `Try` to catch any exceptions thrown before the actually data access operation.